Alertmanager: refactor, 'save as default/global' functionality, new Email and Slack receiver forms

dtaylor113 opened the pull request:
### Refactored:
- updated alert-manager-yaml-editor to use new `getAlertManagerYAML` method in utils lib
- moved single-use functions out of utils, per @kyoto 's [comment](https://github.com/openshift/console/pull/3108#issuecomment-548269155)
- moved sub-form specific logic to seperate subform files
- updated ReceiverBaseForm, , per @kyoto 's [comment](https://github.com/openshift/console/pull/3108#issuecomment-548721461), to:
  - clearly show INIT value handling
  - generically handle subforms via subFormFactory
  - implemended useFormStore custom hook for generic storing of form values

![image](https://user-images.githubusercontent.com/12733153/70275538-548afd80-177c-11ea-96c5-071a49ce9bfc.png)

### Save as Default/global functionality
![image](https://user-images.githubusercontent.com/12733153/70558145-4fe69080-1b52-11ea-932c-6796ddacd11b.png)
**New Integration Test**
![image](https://user-images.githubusercontent.com/12733153/70557417-bc609000-1b50-11ea-8f47-5115306e1be4.png)

### Email Receiver
![image](https://user-images.githubusercontent.com/12733153/70556369-c5506200-1b4e-11ea-92c9-1920fc245a00.png)

### Slack Receiver
![image](https://user-images.githubusercontent.com/12733153/70556398-d305e780-1b4e-11ea-90b1-7a24d2cffad3.png)

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

kyoto commented on frontend/public/components/monitoring/alert-manager-yaml-editor.tsx:
```diff
@@ -63,25 +58,41 @@ const AlertManagerYAMLEditor: React.FC<AlertManagerYAMLEditorProps> = ({ obj })
-        )}
-        {successMsg && <Alert isInline className="co-alert" variant="success" title={successMsg} />}
-      </EditAlertmanagerYAML>
+      {loadErrorMsg && (
```
Probably better as an early return.

```
if (loadErrorMsg) {
  return ...
}
```

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/alert-manager-yaml-editor.tsx:
```diff
@@ -21,16 +20,12 @@ const EditAlertmanagerYAML = (props) => (
+  const [errorMsg, setErrorMsg] = React.useState('');
+  const [loadErrorMsg, setloadErrorMsg] = React.useState('');
+  const [successMsg, setSuccessMsg] = React.useState('');
+  const [alertManagerYamlStr] = React.useState(
```
Looks like `React.useState` was added by mistake here?

dtaylor113 replied:
yes, sorry, was `useState` but no longer needed

dtaylor113 replied:
fixed

kyoto commented on frontend/integration-tests/protractor.conf.ts:
```diff
@@ -228,7 +228,8 @@ export const checkErrors = async () =>
 
-export const firstElementByTestID = (id: string) => $(`[data-test-id=${id}]`);
+export const firstElementByTestID = (id: string) =>
+  element.all(by.css(`[data-test-id=${id}]`)).first();
```
Could you explain why this needed to change?

dtaylor113 replied:
Yes, the integration tests was throwing a 'multiple instances detected but first one will be used' when I was using just the `$`

kyoto replied:
👍

spadgett replied:
I'm not sure we should be suppressing that warning. In a lot of cases, it's not expected, and you'd want to know. I'd suggest separate helpers when we expect there to be more than one. Something like (untested)

```ts
export const elementByTestID = (id: string) => $(`[data-test-id=${id}]`);
export const firstElementByTestID = (id: string) => $$(`[data-test-id=${id}]`)[0];
```

spadgett replied:
I'd even suggest

```ts
export const allElementsByTestID = (id: string) => $$(`[data-test-id=${id}]`);
```

Then call it like `allElementsByTestID('my-id')[0]`. This makes it very clear what's happening.

Is there a way we can make the test ID unique on the page, though, to avoid this problem?

dtaylor113 replied:
I like `firstElementByTestID = (id: string) => $$(`[data-test-id=${id}]`)[0];`

dtaylor113 replied:
I'm using `firstElementByTestID` to get the first drop-down menu item, probably not the best if menu order changes.  I'll look into being able to pass in `data-test-id` to the dropdown component.

spadgett replied:
> I'm using `firstElementByTestID` to get the first drop-down menu item, probably not the best if menu order changes. I'll look into being able to pass in `data-test-id` to the dropdown component.

+1, this is fragile. I would add a specific `data-test-id`

dtaylor113 replied:
Ok, was able to replace it with:
`await firstElementByTestID('dropdown-menu-pagerduty_configs').click();` 
after updating [this](https://github.com/openshift/console/pull/3698/commits/08ec3c02d2937270338ffd5685a68923621f4f99#diff-447b013466c334bed9fb476dc38fc194R126) and [this](https://github.com/openshift/console/pull/3698/commits/08ec3c02d2937270338ffd5685a68923621f4f99#diff-73ded5241d763da982ede2d7c59d3c94R10).  Also reverted changes to `firstElementByTestID`.

dtaylor113 commented on frontend/public/components/monitoring/receiver-forms/email-receiver-form.tsx:
```diff
@@ -0,0 +1,265 @@
+        config,
+        saveAsGlobal ? `smtp_${propName}` : propName,
+        formValue,
+        // TODO: does auth_secret need to be encoded?
```
@kyoto Do you know if we need to Base64.encode auth_secret which is a a CRAM-MDS value type.

dtaylor113 replied:
@spadgett, any thoughts on ^^?  -thanks

dtaylor113 commented:
/retest

kyoto commented on frontend/public/components/monitoring/receiver-forms/webhook-receiver-form.tsx:
```diff
@@ -0,0 +1,40 @@
+);
+
+export const getInitialValues = (globals, receiverConfig) => {
+  return _.isEmpty(receiverConfig)
```
You could use `_.get`'s default value parameter like `return {webhookUrl: _.get(receiverConfig, 'url', '')}`

dtaylor113 replied:
your right!  I did notice that at one point but missed fixing it -thanks

kyoto commented on frontend/public/components/monitoring/alert-manager-yaml-editor.tsx:
```diff
@@ -21,15 +20,11 @@ const EditAlertmanagerYAML = (props) => (
+  const [errorMsg, setErrorMsg] = React.useState('');
+  const [loadErrorMsg, setloadErrorMsg] = React.useState('');
+  const [successMsg, setSuccessMsg] = React.useState('');
+  const alertManagerYamlStr = _.isEmpty(loadErrorMsg)
```
I guess you could drop the `_.isEmpty(loadErrorMsg)` check if you move this after the `if (loadErrorMsg) { ... }` block.

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/alert-manager-utils.tsx:
```diff
@@ -4,28 +4,45 @@ import { Base64 } from 'js-base64';
+  return yaml;
+};
+
+export const getAlertManagerYAML = (secret: K8sResourceKind, setErrorMsg): string => {
```
`getAlertManagerYAML` is an alias for `decodeAlertManagerYaml`, so maybe just export `decodeAlertManagerYaml` instead?

dtaylor113 replied:
I like the symmetry of  `getAlertManagerYAML` and `getAlertManagerConfig`.  One gives yaml string, the other gives config obj.

kyoto replied:
OK. You could keep the name `getAlertManagerYAML`, but still merge the two functions. I think this function actually just being an alias for another is a bit confusing.

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/receiver-forms/email-receiver-form.tsx:
```diff
@@ -0,0 +1,265 @@
+
+export const getInitialValues = (globals, receiverConfig) => {
+  const getInitialValue = (propName: string) =>
+    _.get(receiverConfig, propName) || _.get(globals, `smtp_${propName}`) || '';
```
I think we know that `globals` is always an object, so we can drop some of the `_.get(globals, ...)` here and in the other `getInitialValues` functions.

dtaylor113 replied:
The `smtp_...` field may not be defined in the globals object, I thought using `_.get` returns false as opposed to getting run time error of 'property not defined' if I used `globals.smtp_auth_identity` when `smtp_auth_identity` isn't defined.

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,491 @@
+export const equalOrEmpty = (formValue, globalValue) =>
+  formValue === globalValue || _.isEmpty(formValue);
+export const notEqualAndNotEmpty = (formValue, globalValue) =>
+  formValue !== globalValue && !_.isEmpty(formValue);
```
I realize these eliminate some duplicate code, but I think they're too low level. I'd say either replace them with something more semantic, or just write out each condition in full where it's needed.

Do we need to call `_.isEmpty` here or should it be `formValue === undefined`?

dtaylor113 replied:
Could you pls provide example of 'more semantic'? I use these functions in three receiver forms, I believe it is good coding practice to reduce duplicate code (DRY principle), plus it is very clear what these functions do -makes the code easier to follow IMO.

dtaylor113 replied:
> Do we need to call _.isEmpty here or should it be formValue === undefined?

`_.isEmpty` because formValue may have an initial value of ''.

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,491 @@
+  if (_.isEmpty(errorMsg)) {
+    if (!alertManagerBaseURL) {
+      setErrorMsg(`Error alertManagerBaseURL not set`);
+      return;
```
This is an early return for the component, so the error message will not be displayed.

dtaylor113 replied:
It does get displayed:
![image](https://user-images.githubusercontent.com/12733153/70637208-db205e80-1c04-11ea-869f-900210590d9c.png)
via [here](https://github.com/openshift/console/blob/446148a9eebea9b075560368bb383788a8ddaf44/frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx#L440)
I believe setting the state reloads component.

kyoto replied:
Thanks for the explanation. In cases like this, I think it's more typically to return the error message directly rather than trigger a state change, but I'd like to hear @spadgett's opinion on this.

dtaylor113 replied:
yes, but some of the errors are asynchronous, so just using the same error mechanize.

dtaylor113 replied:
Changed to `useEffect` and was able to remove some of the logic.

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,491 @@
+    }
+  }
+
+  return inProgress ? (
```
Using early returns like `if (inProgress) { ... }` or `if (errorMsg) { ... }` (or both) better matches the coding style we have elsewhere in Console.

dtaylor113 replied:
oh, ok.  When I was working on chargeback reports, metering, those code reviews advocated the conditional operator (condition ? true : false)

kyoto replied:
Well, I think one `? :` is fine as long as the blocks are not too big. In this case there's 3 blocks, so if you make one an early return, I think that is fine.

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,491 @@
+  const [inProgress, setInProgress] = React.useState(false);
+  const [errorMsg, setErrorMsg] = React.useState('');
+
+  if (_.isEmpty(errorMsg)) {
```
I think there's too much use of `_.isEmpty` and `_.get` in general in these files. It's often unnecessary and it can make it less obvious what the expected data type is.

For example, in this case you could use `const [errorMsg, setErrorMsg] = React.useState()` (default to `undefined`), then you can use `if (!errorMsg)` instead.

I think there are a number of other places in this PR where the lodash funtions could be dropped.

dtaylor113 replied:
Ok, I'll take a pass at reducing these

dtaylor113 commented:
Hi @kyoto, saved latest changes in separate commit.

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,493 @@
+        setErrorMsg('');
+        setInProgress(false);
+      });
+      return;
```
This is a return statement for `render()`, so should be `return null` (https://reactjs.org/docs/conditional-rendering.html#preventing-component-from-rendering)

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,493 @@
+  if (_.isEmpty(errorMsg)) {
+    if (!alertManagerBaseURL) {
+      setErrorMsg(`Error alertManagerBaseURL not set`);
+      return;
```
This is a return statement for `render()`, so should be `return null;` (https://reactjs.org/docs/conditional-rendering.html#preventing-component-from-rendering)

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,493 @@
+        const origAlertMngrConfigStr = _.get(data, 'config.original', {});
+        if (_.isEmpty(origAlertMngrConfigStr)) {
+          setErrorMsg('alertmanager.v2.status.config.original not found.');
+          setInProgress(false);
```
Looks like `then()` should always set `setInProgress(false)`, so you could just call that once at the beginning of the  function.

dtaylor113 replied:
fixed

kyoto commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -0,0 +1,493 @@
+      setErrorMsg(`Error alertManagerBaseURL not set`);
+      return;
+    }
+    if (!inProgress && _.isEmpty(alertManagerGlobals)) {
```
This should probably be changed to `useEffect`

dtaylor113 replied:
Thanks, yes, much better. Changed to `useEffect` and was able to remove some of the logic.

dtaylor113 commented:
Need to put this PR on hold, just realized that if you edit a receiver via the form, it will replace the receiver with just values from the form.  If there are other receiver values such as 'send_resolved' which may be added via YAML Editor, they would be removed!  Working on a fix/updated

dtaylor113 commented:
Hi @benjaminapetersen, I believe I have addressed all review comments.  I have a follow up PR which adds advanced configuration fields to the 4 receiver forms -thanks

dtaylor113 commented:
Hi @spadgett, @benjaminapetersen, and @kyoto,
Please review this PR when you get a moment.   Note: there is a follow up PR which continues along this path, so you will get another opportunity to review Alertmanager config code.  -thanks

spadgett commented:
@dtaylor113 I'm still digging into this, but I'm a little uncertain about the subform approach. I think what we did for IDPs is a little easier to follow. We made each IDP a separate form and created shared components and utilities for anything that was common across each form. It scaled well while still avoiding duplication. Here's an example:

https://github.com/openshift/console/blob/master/frontend/public/components/cluster-settings/htpasswd-idp-form.tsx

I find this easier to reason about and less error-prone.

spadgett commented:
I guess the difference here is that it's one form with a dropdown to change types.

@cshinn In other contexts, we use this pattern. Should we consider it here?

![Secrets · Red Hat OpenShift Container Platform 2020-01-10 13-06-30](https://user-images.githubusercontent.com/1167259/72175381-0932c300-33aa-11ea-8dad-a7c552a5d1fb.png)

![cluster · Details · Red Hat OpenShift Container Platform 2020-01-10 13-05-42](https://user-images.githubusercontent.com/1167259/72175352-fc15d400-33a9-11ea-889c-06b0a4962c7f.png)

dtaylor113 commented:
> I guess the difference here is that it's one form with a dropdown to change types.
> @cshinn In other contexts, we use this pattern. Should we consider it here?

@spadgett 
I hope for a follow on PR as this has been the design from the beginning and I hope to get a follow on PR in for adv. config fields which would complete Alertmanager configuration  v1.  I can see v2 implementing any design changes.

dtaylor113 commented:
Closing this large PR to break out into smaller PRs.
